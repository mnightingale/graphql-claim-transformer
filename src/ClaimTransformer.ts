import {
  ArgumentNode,
  DirectiveNode,
  FieldDefinitionNode,
  InterfaceTypeDefinitionNode,
  ObjectTypeDefinitionNode,
  valueFromASTUntyped,
} from 'graphql'
import {
  compoundExpression,
  ifElse,
  not,
  printBlock,
  qref,
  raw,
  ref,
  set,
} from 'graphql-mapping-template'
import { getBaseType, ResolverResourceIDs } from 'graphql-transformer-common'
import {
  gql,
  InvalidDirectiveError,
  Transformer,
  TransformerContext,
} from 'graphql-transformer-core'

export class ClaimTransformer extends Transformer {
  constructor() {
    super(
      'ClaimTransformer',
      gql`
        directive @claim(name: String!) on FIELD_DEFINITION
      `,
    )
  }

  field = (
    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
    definition: FieldDefinitionNode,
    directive: DirectiveNode,
    acc: TransformerContext,
  ): void => {
    if (!['String', 'Int'].includes(getBaseType(definition.type))) {
      throw new InvalidDirectiveError(
        'Directive "claim" must be used only on String or Int type fields.',
      )
    }

    const isArg = (s: string) => (arg: ArgumentNode) => arg.name.value === s
    const getArg = (arg: string, dflt?: any) => {
      const argument = directive.arguments!.find(isArg(arg))
      return argument ? valueFromASTUntyped(argument.value) : dflt
    }

    const claim = getArg('name')
    const field = definition.name.value
    const typeName = parent.name.value

    this.augmentMutation(
      acc,
      ResolverResourceIDs.DynamoDBCreateResolverResourceID(typeName),
      field,
      claim,
    )
    this.augmentMutation(
      acc,
      ResolverResourceIDs.DynamoDBUpdateResolverResourceID(typeName),
      field,
      claim,
    )
  }

  private augmentMutation(
    ctx: TransformerContext,
    mutationResolverLogicalId: string,
    field: string,
    claim: string,
  ) {
    const snippet = printBlock(`Setting "${field}" to claim "${claim}"`)(
      compoundExpression([
        set(ref('claimValue'), raw(`$ctx.identity.claims.get("${claim}")`)),
        ifElse(
          not(raw(`$util.isNullOrEmpty($claimValue)`)),
          qref(`$ctx.args.input.put("${field}", $claimValue)`), // Use the value from claims
          qref(`$ctx.args.input.remove("${field}")`), // Remove any user supplied value
        ),
      ]),
    )
    const resolver = ctx.getResource(mutationResolverLogicalId)
    if (resolver) {
      resolver.Properties!.RequestMappingTemplate =
        snippet + '\n\n' + resolver.Properties!.RequestMappingTemplate
      ctx.setResource(mutationResolverLogicalId, resolver)
    }
  }
}
