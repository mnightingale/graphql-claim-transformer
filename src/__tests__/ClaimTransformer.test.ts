// @ts-ignore
import { AppSyncTransformer } from 'graphql-appsync-transformer'
import { DynamoDBModelTransformer } from 'graphql-dynamodb-transformer'
import { GraphQLTransform } from 'graphql-transformer-core'
import ClaimTransformer from '../index'

const featureFlags = {
  getBoolean: jest.fn().mockImplementation((name, defaultValue) => {
    if (name === 'improvePluralization') {
      return true
    }
    return
  }),
  getNumber: jest.fn(),
  getObject: jest.fn(),
  getString: jest.fn(),
}

const transformer = new GraphQLTransform({
  featureFlags,
  transformers: [
    new AppSyncTransformer(),
    new DynamoDBModelTransformer(),
    new ClaimTransformer(),
  ],
})

test('@claim directive can be used on fields', () => {
  const schema = `
    type Todo @model {
      id: ID!
      claim: String @claim(name: "foobar")
    }
  `
  expect(() => transformer.transform(schema)).not.toThrow()
})

test('@claim directive can not be used on types', () => {
  const schema = `
    type Todo @model @claim {
      id: ID!
    }
  `
  expect(() => transformer.transform(schema)).toThrowError(
    'Directive "claim" may not be used on OBJECT.',
  )
})

test('@claim directive can not be used on fields other than String and Int', () => {
  const schema = `
    type Todo @model {
      id: ID!
      claim: AWSDateTime @claim(name: "foobar")
    }
  `
  expect(() => transformer.transform(schema)).toThrowError(
    'Directive "claim" must be used only on String or Int type fields.',
  )
})

test('@claim directive can be used on fields with String type', () => {
  const schema = `
    type Todo @model {
      id: ID!
      claim: String @claim(name: "foobar")
    }
  `
  expect(() => transformer.transform(schema)).not.toThrow()
})

test('@claim directive can be used on fields with Int type', () => {
  const schema = `
    type Todo @model {
      id: ID!
      claim: Int @claim(name: "foobar")
    }
  `
  expect(() => transformer.transform(schema)).not.toThrow()
})

test('@claim generates valid resovler code', () => {
  const schema = `
    type Todo @model {
      id: ID!
      foo: String @claim(name: "bar")
    }
  `
  const out = transformer.transform(schema)
  expect(out).toBeDefined()
  expect(out.resolvers).toMatchSnapshot()
})
