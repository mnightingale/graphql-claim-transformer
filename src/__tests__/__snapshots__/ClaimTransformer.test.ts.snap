// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@claim generates valid resovler code 1`] = `
Object {
  "Mutation.createTodo.req.vtl": "## [Start] Set default values. **
$util.qr($context.args.input.put(\\"id\\", $util.defaultIfNull($ctx.args.input.id, $util.autoId())))
#set( $createdAt = $util.time.nowISO8601() )
## Automatically set the createdAt timestamp. **
$util.qr($context.args.input.put(\\"createdAt\\", $util.defaultIfNull($ctx.args.input.createdAt, $createdAt)))
## Automatically set the updatedAt timestamp. **
$util.qr($context.args.input.put(\\"updatedAt\\", $util.defaultIfNull($ctx.args.input.updatedAt, $createdAt)))
## [End] Set default values. **
## [Start] Setting \\"foo\\" to claim \\"bar\\". **
#if( !$util.isNullOrEmpty($ctx.identity.claims.get(\\"bar\\")) )
  $util.qr($ctx.args.input.put(\\"foo\\", \\"$ctx.identity.claims.get(\\"bar\\")\\"))
#end
## [End] Setting \\"foo\\" to claim \\"bar\\". **

## [Start] Prepare DynamoDB PutItem Request. **
$util.qr($context.args.input.put(\\"__typename\\", \\"Todo\\"))
#if( $modelObjectKey )
  #set( $condition = {
  \\"expression\\": \\"\\",
  \\"expressionNames\\": {},
  \\"expressionValues\\": {}
} )
  #foreach( $entry in $modelObjectKey.entrySet() )
    #if( $velocityCount == 1 )
      $util.qr($condition.put(\\"expression\\", \\"attribute_not_exists(#keyCondition$velocityCount)\\"))
    #else
      $util.qr($condition.put(\\"expression\\", \\"$condition.expression AND attribute_not_exists(#keyCondition$velocityCount)\\"))
    #end
    $util.qr($condition.expressionNames.put(\\"#keyCondition$velocityCount\\", \\"$entry.key\\"))
  #end
#else
  #set( $condition = {
  \\"expression\\": \\"attribute_not_exists(#id)\\",
  \\"expressionNames\\": {
      \\"#id\\": \\"id\\"
  },
  \\"expressionValues\\": {}
} )
#end
#if( $context.args.condition )
  #set( $condition.expressionValues = {} )
  #set( $conditionFilterExpressions = $util.parseJson($util.transform.toDynamoDBConditionExpression($context.args.condition)) )
  $util.qr($condition.put(\\"expression\\", \\"($condition.expression) AND $conditionFilterExpressions.expression\\"))
  $util.qr($condition.expressionNames.putAll($conditionFilterExpressions.expressionNames))
  $util.qr($condition.expressionValues.putAll($conditionFilterExpressions.expressionValues))
#end
#if( $condition.expressionValues && $condition.expressionValues.size() == 0 )
  #set( $condition = {
  \\"expression\\": $condition.expression,
  \\"expressionNames\\": $condition.expressionNames
} )
#end
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"PutItem\\",
  \\"key\\": #if( $modelObjectKey ) $util.toJson($modelObjectKey) #else {
  \\"id\\":   $util.dynamodb.toDynamoDBJson($ctx.args.input.id)
} #end,
  \\"attributeValues\\": $util.dynamodb.toMapValuesJson($context.args.input),
  \\"condition\\": $util.toJson($condition)
}
## [End] Prepare DynamoDB PutItem Request. **",
  "Mutation.createTodo.res.vtl": "#if( $ctx.error )
$util.error($ctx.error.message, $ctx.error.type)
#else
$util.toJson($ctx.result)
#end",
  "Mutation.deleteTodo.req.vtl": "#if( $authCondition )
  #set( $condition = $authCondition )
  #if( $modelObjectKey )
    #foreach( $entry in $modelObjectKey.entrySet() )
      $util.qr($condition.put(\\"expression\\", \\"$condition.expression AND attribute_exists(#keyCondition$velocityCount)\\"))
      $util.qr($condition.expressionNames.put(\\"#keyCondition$velocityCount\\", \\"$entry.key\\"))
    #end
  #else
    $util.qr($condition.put(\\"expression\\", \\"$condition.expression AND attribute_exists(#id)\\"))
    $util.qr($condition.expressionNames.put(\\"#id\\", \\"id\\"))
  #end
#else
  #if( $modelObjectKey )
    #set( $condition = {
  \\"expression\\": \\"\\",
  \\"expressionNames\\": {},
  \\"expressionValues\\": {}
} )
    #foreach( $entry in $modelObjectKey.entrySet() )
      #if( $velocityCount == 1 )
        $util.qr($condition.put(\\"expression\\", \\"attribute_exists(#keyCondition$velocityCount)\\"))
      #else
        $util.qr($condition.put(\\"expression\\", \\"$condition.expression AND attribute_exists(#keyCondition$velocityCount)\\"))
      #end
      $util.qr($condition.expressionNames.put(\\"#keyCondition$velocityCount\\", \\"$entry.key\\"))
    #end
  #else
    #set( $condition = {
  \\"expression\\": \\"attribute_exists(#id)\\",
  \\"expressionNames\\": {
      \\"#id\\": \\"id\\"
  },
  \\"expressionValues\\": {}
} )
  #end
#end
#if( $versionedCondition )
  $util.qr($condition.put(\\"expression\\", \\"($condition.expression) AND $versionedCondition.expression\\"))
  $util.qr($condition.expressionNames.putAll($versionedCondition.expressionNames))
  #set( $expressionValues = $util.defaultIfNull($condition.expressionValues, {}) )
  $util.qr($expressionValues.putAll($versionedCondition.expressionValues))
  #set( $condition.expressionValues = $expressionValues )
#end
#if( $context.args.condition )
  #set( $conditionFilterExpressions = $util.parseJson($util.transform.toDynamoDBConditionExpression($context.args.condition)) )
  $util.qr($condition.put(\\"expression\\", \\"($condition.expression) AND $conditionFilterExpressions.expression\\"))
  $util.qr($condition.expressionNames.putAll($conditionFilterExpressions.expressionNames))
  #set( $conditionExpressionValues = $util.defaultIfNull($condition.expressionValues, {}) )
  $util.qr($conditionExpressionValues.putAll($conditionFilterExpressions.expressionValues))
  #set( $condition.expressionValues = $conditionExpressionValues )
  $util.qr($condition.expressionValues.putAll($conditionFilterExpressions.expressionValues))
#end
#if( $condition.expressionValues && $condition.expressionValues.size() == 0 )
  #set( $condition = {
  \\"expression\\": $condition.expression,
  \\"expressionNames\\": $condition.expressionNames
} )
#end
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"DeleteItem\\",
  \\"key\\": #if( $modelObjectKey ) $util.toJson($modelObjectKey) #else {
  \\"id\\": $util.dynamodb.toDynamoDBJson($ctx.args.input.id)
} #end,
  \\"condition\\": $util.toJson($condition)
}",
  "Mutation.deleteTodo.res.vtl": "#if( $ctx.error )
$util.error($ctx.error.message, $ctx.error.type)
#else
$util.toJson($ctx.result)
#end",
  "Mutation.updateTodo.req.vtl": "## [Start] Setting \\"foo\\" to claim \\"bar\\". **
#if( !$util.isNullOrEmpty($ctx.identity.claims.get(\\"bar\\")) )
  $util.qr($ctx.args.input.put(\\"foo\\", \\"$ctx.identity.claims.get(\\"bar\\")\\"))
#end
## [End] Setting \\"foo\\" to claim \\"bar\\". **

#set( $optionalNonNullableFields = [] )
#foreach( $field in $optionalNonNullableFields )
  #if( $context.arguments.input.keySet().contains($field) && $util.isNull($context.args.input.get($field)) )
$util.error(\\"An argument you marked as Non-Null is set to Null in the query or the body of your request.\\")
  #end
#end
#if( $authCondition && $authCondition.expression != \\"\\" )
  #set( $condition = $authCondition )
  #if( $modelObjectKey )
    #foreach( $entry in $modelObjectKey.entrySet() )
      $util.qr($condition.put(\\"expression\\", \\"$condition.expression AND attribute_exists(#keyCondition$velocityCount)\\"))
      $util.qr($condition.expressionNames.put(\\"#keyCondition$velocityCount\\", \\"$entry.key\\"))
    #end
  #else
    $util.qr($condition.put(\\"expression\\", \\"$condition.expression AND attribute_exists(#id)\\"))
    $util.qr($condition.expressionNames.put(\\"#id\\", \\"id\\"))
  #end
#else
  #if( $modelObjectKey )
    #set( $condition = {
  \\"expression\\": \\"\\",
  \\"expressionNames\\": {},
  \\"expressionValues\\": {}
} )
    #foreach( $entry in $modelObjectKey.entrySet() )
      #if( $velocityCount == 1 )
        $util.qr($condition.put(\\"expression\\", \\"attribute_exists(#keyCondition$velocityCount)\\"))
      #else
        $util.qr($condition.put(\\"expression\\", \\"$condition.expression AND attribute_exists(#keyCondition$velocityCount)\\"))
      #end
      $util.qr($condition.expressionNames.put(\\"#keyCondition$velocityCount\\", \\"$entry.key\\"))
    #end
  #else
    #set( $condition = {
  \\"expression\\": \\"attribute_exists(#id)\\",
  \\"expressionNames\\": {
      \\"#id\\": \\"id\\"
  },
  \\"expressionValues\\": {}
} )
  #end
#end
## Automatically set the updatedAt timestamp. **
$util.qr($context.args.input.put(\\"updatedAt\\", $util.defaultIfNull($ctx.args.input.updatedAt, $util.time.nowISO8601())))
$util.qr($context.args.input.put(\\"__typename\\", \\"Todo\\"))
## Update condition if type is @versioned **
#if( $versionedCondition )
  $util.qr($condition.put(\\"expression\\", \\"($condition.expression) AND $versionedCondition.expression\\"))
  $util.qr($condition.expressionNames.putAll($versionedCondition.expressionNames))
  $util.qr($condition.expressionValues.putAll($versionedCondition.expressionValues))
#end
#if( $context.args.condition )
  #set( $conditionFilterExpressions = $util.parseJson($util.transform.toDynamoDBConditionExpression($context.args.condition)) )
  $util.qr($condition.put(\\"expression\\", \\"($condition.expression) AND $conditionFilterExpressions.expression\\"))
  $util.qr($condition.expressionNames.putAll($conditionFilterExpressions.expressionNames))
  $util.qr($condition.expressionValues.putAll($conditionFilterExpressions.expressionValues))
#end
#if( $condition.expressionValues && $condition.expressionValues.size() == 0 )
  #set( $condition = {
  \\"expression\\": $condition.expression,
  \\"expressionNames\\": $condition.expressionNames
} )
#end
#set( $expNames = {} )
#set( $expValues = {} )
#set( $expSet = {} )
#set( $expAdd = {} )
#set( $expRemove = [] )
#if( $modelObjectKey )
  #set( $keyFields = [] )
  #foreach( $entry in $modelObjectKey.entrySet() )
    $util.qr($keyFields.add(\\"$entry.key\\"))
  #end
#else
  #set( $keyFields = [\\"id\\"] )
#end
#foreach( $entry in $util.map.copyAndRemoveAllKeys($context.args.input, $keyFields).entrySet() )
  #if( !$util.isNull($dynamodbNameOverrideMap) && $dynamodbNameOverrideMap.containsKey(\\"$entry.key\\") )
    #set( $entryKeyAttributeName = $dynamodbNameOverrideMap.get(\\"$entry.key\\") )
  #else
    #set( $entryKeyAttributeName = $entry.key )
  #end
  #if( $util.isNull($entry.value) )
    #set( $discard = $expRemove.add(\\"#$entryKeyAttributeName\\") )
    $util.qr($expNames.put(\\"#$entryKeyAttributeName\\", \\"$entry.key\\"))
  #else
    $util.qr($expSet.put(\\"#$entryKeyAttributeName\\", \\":$entryKeyAttributeName\\"))
    $util.qr($expNames.put(\\"#$entryKeyAttributeName\\", \\"$entry.key\\"))
    $util.qr($expValues.put(\\":$entryKeyAttributeName\\", $util.dynamodb.toDynamoDB($entry.value)))
  #end
#end
#set( $expression = \\"\\" )
#if( !$expSet.isEmpty() )
  #set( $expression = \\"SET\\" )
  #foreach( $entry in $expSet.entrySet() )
    #set( $expression = \\"$expression $entry.key = $entry.value\\" )
    #if( $foreach.hasNext() )
      #set( $expression = \\"$expression,\\" )
    #end
  #end
#end
#if( !$expAdd.isEmpty() )
  #set( $expression = \\"$expression ADD\\" )
  #foreach( $entry in $expAdd.entrySet() )
    #set( $expression = \\"$expression $entry.key $entry.value\\" )
    #if( $foreach.hasNext() )
      #set( $expression = \\"$expression,\\" )
    #end
  #end
#end
#if( !$expRemove.isEmpty() )
  #set( $expression = \\"$expression REMOVE\\" )
  #foreach( $entry in $expRemove )
    #set( $expression = \\"$expression $entry\\" )
    #if( $foreach.hasNext() )
      #set( $expression = \\"$expression,\\" )
    #end
  #end
#end
#set( $update = {} )
$util.qr($update.put(\\"expression\\", \\"$expression\\"))
#if( !$expNames.isEmpty() )
  $util.qr($update.put(\\"expressionNames\\", $expNames))
#end
#if( !$expValues.isEmpty() )
  $util.qr($update.put(\\"expressionValues\\", $expValues))
#end
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"UpdateItem\\",
  \\"key\\": #if( $modelObjectKey ) $util.toJson($modelObjectKey) #else {
  \\"id\\": {
      \\"S\\": $util.toJson($context.args.input.id)
  }
} #end,
  \\"update\\": $util.toJson($update),
  \\"condition\\": $util.toJson($condition)
}",
  "Mutation.updateTodo.res.vtl": "#if( $ctx.error )
$util.error($ctx.error.message, $ctx.error.type)
#else
$util.toJson($ctx.result)
#end",
  "Query.getTodo.req.vtl": "{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"GetItem\\",
  \\"key\\": #if( $modelObjectKey ) $util.toJson($modelObjectKey) #else {
  \\"id\\": $util.dynamodb.toDynamoDBJson($ctx.args.id)
} #end
}",
  "Query.getTodo.res.vtl": "#if( $ctx.error )
$util.error($ctx.error.message, $ctx.error.type)
#else
$util.toJson($ctx.result)
#end",
  "Query.listTodos.req.vtl": "#set( $limit = $util.defaultIfNull($context.args.limit, 100) )
#set( $ListRequest = {
  \\"version\\": \\"2018-05-29\\",
  \\"limit\\": $limit
} )
#if( $context.args.nextToken )
  #set( $ListRequest.nextToken = $context.args.nextToken )
#end
#if( $context.args.filter )
  #set( $ListRequest.filter = $util.parseJson(\\"$util.transform.toDynamoDBFilterExpression($ctx.args.filter)\\") )
#end
#if( !$util.isNull($modelQueryExpression)
                        && !$util.isNullOrEmpty($modelQueryExpression.expression) )
  $util.qr($ListRequest.put(\\"operation\\", \\"Query\\"))
  $util.qr($ListRequest.put(\\"query\\", $modelQueryExpression))
  #if( !$util.isNull($ctx.args.sortDirection) && $ctx.args.sortDirection == \\"DESC\\" )
    #set( $ListRequest.scanIndexForward = false )
  #else
    #set( $ListRequest.scanIndexForward = true )
  #end
#else
  $util.qr($ListRequest.put(\\"operation\\", \\"Scan\\"))
#end
$util.toJson($ListRequest)",
  "Query.listTodos.res.vtl": "#if( $ctx.error )
$util.error($ctx.error.message, $ctx.error.type)
#else
$util.toJson($ctx.result)
#end",
}
`;
